File: packages/migration-cli/templates/skills/business-logic-mapping/examples/model-example.java
// =============================================================================
// Java EE JPA Entity to Spring Boot JPA Entity Conversion Example
// =============================================================================

// -----------------------------------------------------------------------------
// BEFORE: Java EE JPA Entity (com/example/models/Product.java)
// -----------------------------------------------------------------------------
/*
package com.example.models;

import javax.persistence.*;
import javax.validation.constraints.*;
import java.math.BigDecimal;
import java.util.Date;
import java.util.HashSet;
import java.util.Set;

@Entity
@Table(name = "products")
@NamedQueries({
    @NamedQuery(name = "Product.findAll", 
                query = "SELECT p FROM Product p"),
    @NamedQuery(name = "Product.findByCategory", 
                query = "SELECT p FROM Product p WHERE p.category.id = :categoryId"),
    @NamedQuery(name = "Product.findActive", 
                query = "SELECT p FROM Product p WHERE p.active = true")
})
public class Product {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @NotNull
    @Size(max = 255)
    @Column(nullable = false)
    private String name;

    @NotNull
    @Size(max = 255)
    @Column(nullable = false, unique = true)
    private String slug;

    @Column(columnDefinition = "TEXT")
    private String description;

    @NotNull
    @DecimalMin("0.00")
    @Column(precision = 18, scale = 2, nullable = false)
    private BigDecimal price;

    @Column(name = "is_active")
    private boolean active = true;

    @Column(name = "stock_quantity")
    private int stockQuantity = 0;

    @Temporal(TemporalType.TIMESTAMP)
    @Column(name = "created_at")
    private Date createdAt;

    @Temporal(TemporalType.TIMESTAMP)
    @Column(name = "updated_at")
    private Date updatedAt;

    @Temporal(TemporalType.TIMESTAMP)
    @Column(name = "deleted_at")
    private Date deletedAt;

    // Relationships
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "category_id")
    private Category category;

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL, orphanRemoval = true)
    private Set<Review> reviews = new HashSet<>();

    @ManyToMany
    @JoinTable(
        name = "product_tags",
        joinColumns = @JoinColumn(name = "product_id"),
        inverseJoinColumns = @JoinColumn(name = "tag_id")
    )
    private Set<Tag> tags = new HashSet<>();

    // Lifecycle callbacks
    @PrePersist
    protected void onCreate() {
        createdAt = new Date();
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = new Date();
    }

    // Getters and Setters
    public Long getId() { return id; }
    public void setId(Long id) { this.id = id; }
    
    public String getName() { return name; }
    public void setName(String name) { this.name = name; }
    
    public String getSlug() { return slug; }
    public void setSlug(String slug) { this.slug = slug; }
    
    public String getDescription() { return description; }
    public void setDescription(String description) { this.description = description; }
    
    public BigDecimal getPrice() { return price; }
    public void setPrice(BigDecimal price) { this.price = price; }
    
    public boolean isActive() { return active; }
    public void setActive(boolean active) { this.active = active; }
    
    public int getStockQuantity() { return stockQuantity; }
    public void setStockQuantity(int stockQuantity) { this.stockQuantity = stockQuantity; }
    
    public Date getCreatedAt() { return createdAt; }
    public void setCreatedAt(Date createdAt) { this.createdAt = createdAt; }
    
    public Date getUpdatedAt() { return updatedAt; }
    public void setUpdatedAt(Date updatedAt) { this.updatedAt = updatedAt; }
    
    public Date getDeletedAt() { return deletedAt; }
    public void setDeletedAt(Date deletedAt) { this.deletedAt = deletedAt; }
    
    public Category getCategory() { return category; }
    public void setCategory(Category category) { this.category = category; }
    
    public Set<Review> getReviews() { return reviews; }
    public void setReviews(Set<Review> reviews) { this.reviews = reviews; }
    
    public Set<Tag> getTags() { return tags; }
    public void setTags(Set<Tag> tags) { this.tags = tags; }

    // =========================================================================
    // BUSINESS LOGIC METHODS
    // =========================================================================

    // Business Rule: Calculate formatted price for display
    public String getFormattedPrice() {
        return String.format("$%.2f", price);
    }

    // Business Rule: Check if product is available for purchase
    public boolean isAvailable() {
        return active && stockQuantity > 0 && deletedAt == null;
    }

    // Business Rule: Apply discount with maximum cap
    public BigDecimal applyDiscount(int percentage) {
        // Max discount is 50%
        int effectivePercentage = Math.min(percentage, 50);
        BigDecimal discountMultiplier = BigDecimal.ONE
            .subtract(BigDecimal.valueOf(effectivePercentage).divide(BigDecimal.valueOf(100)));
        return price.multiply(discountMultiplier);
    }

    // Business Rule: Check if product needs restock
    public boolean needsRestock(int threshold) {
        return stockQuantity <= threshold;
    }

    // Business Rule: Add review with validation
    public void addReview(Review review) {
        if (review == null) {
            throw new IllegalArgumentException("Review cannot be null");
        }
        reviews.add(review);
        review.setProduct(this);
    }

    // Business Rule: Calculate average rating
    public double getAverageRating() {
        if (reviews.isEmpty()) {
            return 0.0;
        }
        return reviews.stream()
            .mapToInt(Review::getRating)
            .average()
            .orElse(0.0);
    }
}
*/

// -----------------------------------------------------------------------------
// AFTER: Spring Boot JPA Entity (com/example/model/Product.java)
// -----------------------------------------------------------------------------

package com.example.model;

import jakarta.persistence.*;
import jakarta.validation.constraints.*;
import java.math.BigDecimal;
import java.math.RoundingMode;
import java.time.LocalDateTime;
import java.util.HashSet;
import java.util.Set;

@Entity
@Table(name = "products", indexes = {
    @Index(name = "idx_product_slug", columnList = "slug", unique = true),
    @Index(name = "idx_product_category", columnList = "category_id"),
    @Index(name = "idx_product_active", columnList = "is_active")
})
public class Product {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @NotNull
    @Size(max = 255)
    @Column(nullable = false)
    private String name;

    @NotNull
    @Size(max = 255)
    @Column(nullable = false, unique = true)
    private String slug;

    @Column(columnDefinition = "TEXT")
    private String description;

    @NotNull
    @DecimalMin("0.00")
    @Column(precision = 18, scale = 2, nullable = false)
    private BigDecimal price;

    @Column(name = "is_active")
    private boolean active = true;

    @Column(name = "stock_quantity")
    private int stockQuantity = 0;

    // Use LocalDateTime instead of Date
    @Column(name = "created_at")
    private LocalDateTime createdAt;

    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @Column(name = "deleted_at")
    private LocalDateTime deletedAt;

    // Relationships (same pattern, jakarta.persistence)
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "category_id")
    private Category category;

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL, orphanRemoval = true)
    private Set<Review> reviews = new HashSet<>();

    @ManyToMany
    @JoinTable(
        name = "product_tags",
        joinColumns = @JoinColumn(name = "product_id"),
        inverseJoinColumns = @JoinColumn(name = "tag_id")
    )
    private Set<Tag> tags = new HashSet<>();

    // Lifecycle callbacks (same pattern)
    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }

    // =========================================================================
    // GETTERS AND SETTERS
    // =========================================================================

    public Long getId() { return id; }
    public void setId(Long id) { this.id = id; }
    
    public String getName() { return name; }
    public void setName(String name) { this.name = name; }
    
    public String getSlug() { return slug; }
    public void setSlug(String slug) { this.slug = slug; }
    
    public String getDescription() { return description; }
    public void setDescription(String description) { this.description = description; }
    
    public BigDecimal getPrice() { return price; }
    public void setPrice(BigDecimal price) { this.price = price; }
    
    public boolean isActive() { return active; }
    public void setActive(boolean active) { this.active = active; }
    
    public int getStockQuantity() { return stockQuantity; }
    public void setStockQuantity(int stockQuantity) { this.stockQuantity = stockQuantity; }
    
    public LocalDateTime getCreatedAt() { return createdAt; }
    public void setCreatedAt(LocalDateTime createdAt) { this.createdAt = createdAt; }
    
    public LocalDateTime getUpdatedAt() { return updatedAt; }
    public void setUpdatedAt(LocalDateTime updatedAt) { this.updatedAt = updatedAt; }
    
    public LocalDateTime getDeletedAt() { return deletedAt; }
    public void setDeletedAt(LocalDateTime deletedAt) { this.deletedAt = deletedAt; }
    
    public Category getCategory() { return category; }
    public void setCategory(Category category) { this.category = category; }
    
    public Set<Review> getReviews() { return reviews; }
    public void setReviews(Set<Review> reviews) { this.reviews = reviews; }
    
    public Set<Tag> getTags() { return tags; }
    public void setTags(Set<Tag> tags) { this.tags = tags; }

    // =========================================================================
    // PRESERVED BUSINESS LOGIC METHODS
    // =========================================================================

    /**
     * PRESERVED: Calculate formatted price for display
     */
    public String getFormattedPrice() {
        return String.format("$%.2f", price);
    }

    /**
     * PRESERVED: Check if product is available for purchase
     * Business Rule: Must be active, in stock, and not deleted
     */
    public boolean isAvailable() {
        return active && stockQuantity > 0 && deletedAt == null;
    }

    /**
     * PRESERVED: Apply discount with maximum cap
     * Business Rule: Maximum discount is 50%
     */
    public BigDecimal applyDiscount(int percentage) {
        int effectivePercentage = Math.min(percentage, 50);
        BigDecimal discountMultiplier = BigDecimal.ONE
            .subtract(BigDecimal.valueOf(effectivePercentage)
                .divide(BigDecimal.valueOf(100), 4, RoundingMode.HALF_UP));
        return price.multiply(discountMultiplier).setScale(2, RoundingMode.HALF_UP);
    }

    /**
     * PRESERVED: Check if product needs restock
     * Business Rule: Needs restock when stock falls to or below threshold
     */
    public boolean needsRestock(int threshold) {
        return stockQuantity <= threshold;
    }

    /**
     * PRESERVED: Add review with validation
     * Business Rule: Review cannot be null, maintains bidirectional relationship
     */
    public void addReview(Review review) {
        if (review == null) {
            throw new IllegalArgumentException("Review cannot be null");
        }
        reviews.add(review);
        review.setProduct(this);
    }

    /**
     * PRESERVED: Calculate average rating from reviews
     * Business Rule: Returns 0.0 if no reviews exist
     */
    public double getAverageRating() {
        if (reviews.isEmpty()) {
            return 0.0;
        }
        return reviews.stream()
            .mapToInt(Review::getRating)
            .average()
            .orElse(0.0);
    }

    /**
     * NEW: Remove review (bidirectional)
     */
    public void removeReview(Review review) {
        reviews.remove(review);
        review.setProduct(null);
    }
}

// -----------------------------------------------------------------------------
// Soft Delete Specification (for global filtering)
// -----------------------------------------------------------------------------

package com.example.repository;

import com.example.model.Product;
import org.springframework.data.jpa.domain.Specification;

public class ProductSpecifications {

    public static Specification<Product> notDeleted() {
        return (root, query, criteriaBuilder) -> 
            criteriaBuilder.isNull(root.get("deletedAt"));
    }

    public static Specification<Product> isActive() {
        return (root, query, criteriaBuilder) -> 
            criteriaBuilder.isTrue(root.get("active"));
    }

    public static Specification<Product> inCategory(Long categoryId) {
        return (root, query, criteriaBuilder) -> 
            criteriaBuilder.equal(root.get("category").get("id"), categoryId);
    }

    public static Specification<Product> priceBetween(BigDecimal min, BigDecimal max) {
        return (root, query, criteriaBuilder) -> 
            criteriaBuilder.between(root.get("price"), min, max);
    }
}

// Usage in service:
// productRepository.findAll(
//     Specification.where(notDeleted())
//         .and(isActive())
//         .and(inCategory(categoryId))
// );

// -----------------------------------------------------------------------------
// KEY CONVERSION NOTES:
// -----------------------------------------------------------------------------
// 1. javax.persistence → jakarta.persistence (Jakarta EE 9+)
// 2. javax.validation → jakarta.validation
// 3. java.util.Date → java.time.LocalDateTime
// 4. @Temporal not needed with LocalDateTime
// 5. @NamedQueries → Spring Data derived methods or @Query
// 6. Business methods PRESERVED exactly as-is
// 7. Same @PrePersist/@PreUpdate callbacks work
// 8. Same relationship mappings work
// 9. Added @Index annotations for performance
// 10. Use Specification pattern for dynamic queries (replaces Criteria API)
// 11. RoundingMode added for precise decimal calculations
// 12. HashSet still works for collections

File: packages/migration-cli/templates/skills/business-logic-mapping/examples/model-example.cs
// =============================================================================
// .NET Framework EF6 to .NET 10 EF Core Entity Conversion Example
// =============================================================================

// -----------------------------------------------------------------------------
// BEFORE: .NET Framework EF6 Entity (Models/Product.cs)
// -----------------------------------------------------------------------------
/*
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace ProjectName.Models
{
    [Table("Products")]
    public class Product
    {
        [Key]
        public int Id { get; set; }

        [Required]
        [StringLength(255)]
        public string Name { get; set; }

        [Required]
        [StringLength(255)]
        public string Slug { get; set; }

        public string Description { get; set; }

        [Column(TypeName = "decimal(18,2)")]
        public decimal Price { get; set; }

        public int CategoryId { get; set; }

        public bool IsActive { get; set; }

        public string MetadataJson { get; set; }

        public DateTime CreatedAt { get; set; }
        public DateTime? UpdatedAt { get; set; }
        public DateTime? DeletedAt { get; set; }

        // Navigation Properties
        [ForeignKey("CategoryId")]
        public virtual Category Category { get; set; }

        public virtual ICollection<Review> Reviews { get; set; }

        public virtual ICollection<Tag> Tags { get; set; }

        public virtual ICollection<Image> Images { get; set; }

        // Constructor
        public Product()
        {
            Reviews = new HashSet<Review>();
            Tags = new HashSet<Tag>();
            Images = new HashSet<Image>();
            IsActive = true;
        }

        // Business Logic Methods
        public string GetFormattedPrice()
        {
            return string.Format("${0:N2}", Price);
        }

        public bool IsAvailable(int stockQuantity)
        {
            return IsActive && stockQuantity > 0;
        }

        public decimal ApplyDiscount(decimal percentage)
        {
            return Price * (1 - percentage / 100);
        }
    }
}

// EF6 DbContext
using System.Data.Entity;

namespace ProjectName.Data
{
    public class ApplicationDbContext : DbContext
    {
        public ApplicationDbContext() : base("DefaultConnection")
        {
            // Disable lazy loading
            Configuration.LazyLoadingEnabled = false;
        }

        public DbSet<Product> Products { get; set; }
        public DbSet<Category> Categories { get; set; }
        public DbSet<Review> Reviews { get; set; }
        public DbSet<Tag> Tags { get; set; }

        protected override void OnModelCreating(DbModelBuilder modelBuilder)
        {
            base.OnModelCreating(modelBuilder);

            // Many-to-many relationship
            modelBuilder.Entity<Product>()
                .HasMany(p => p.Tags)
                .WithMany(t => t.Products)
                .Map(m =>
                {
                    m.ToTable("ProductTags");
                    m.MapLeftKey("ProductId");
                    m.MapRightKey("TagId");
                });

            // Soft delete filter (manual implementation)
            // Note: EF6 doesn't have global query filters
        }

        public override int SaveChanges()
        {
            // Auto-set timestamps
            foreach (var entry in ChangeTracker.Entries<Product>())
            {
                if (entry.State == EntityState.Added)
                {
                    entry.Entity.CreatedAt = DateTime.Now;
                }
                else if (entry.State == EntityState.Modified)
                {
                    entry.Entity.UpdatedAt = DateTime.Now;
                }
            }

            return base.SaveChanges();
        }
    }
}
*/

// -----------------------------------------------------------------------------
// AFTER: .NET 10 EF Core Entity (Models/Entities/Product.cs)
// -----------------------------------------------------------------------------

using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace ProjectName.Models.Entities;

public class Product
{
    public int Id { get; set; }

    [Required]
    [MaxLength(255)]
    public string Name { get; set; } = string.Empty;

    [Required]
    [MaxLength(255)]
    public string Slug { get; set; } = string.Empty;

    public string? Description { get; set; }

    [Column(TypeName = "decimal(18,2)")]
    public decimal Price { get; set; }

    public int CategoryId { get; set; }

    public bool IsActive { get; set; } = true;

    public string? MetadataJson { get; set; }

    public DateTime CreatedAt { get; set; }
    public DateTime? UpdatedAt { get; set; }
    public DateTime? DeletedAt { get; set; } // Soft delete

    // Navigation Properties (same pattern, no virtual keyword needed)
    
    // Foreign key relationship
    public Category Category { get; set; } = null!;

    // One-to-many
    public ICollection<Review> Reviews { get; set; } = new List<Review>();

    // Many-to-many via join entity (EF Core pattern)
    public ICollection<ProductTag> ProductTags { get; set; } = new List<ProductTag>();

    // Polymorphic relationship
    public ICollection<Image> Images { get; set; } = new List<Image>();

    // Computed Property (replaces method for simple formatting)
    [NotMapped]
    public string FormattedPrice => $"${Price:N2}";

    // PRESERVED: Business Logic Methods (same implementation)
    public bool IsAvailable(int stockQuantity) => IsActive && stockQuantity > 0;

    public decimal ApplyDiscount(decimal percentage) => Price * (1 - percentage / 100);
}

// -----------------------------------------------------------------------------
// Join Entity for Many-to-Many (Models/Entities/ProductTag.cs)
// EF Core requires explicit join entity for many-to-many with payload
// -----------------------------------------------------------------------------

namespace ProjectName.Models.Entities;

public class ProductTag
{
    public int ProductId { get; set; }
    public Product Product { get; set; } = null!;

    public int TagId { get; set; }
    public Tag Tag { get; set; } = null!;
    
    // Can add additional properties (e.g., CreatedAt, SortOrder)
}

// -----------------------------------------------------------------------------
// EF Core DbContext (Data/ApplicationDbContext.cs)
// -----------------------------------------------------------------------------

using Microsoft.EntityFrameworkCore;
using ProjectName.Models.Entities;

namespace ProjectName.Data;

public class ApplicationDbContext : DbContext
{
    public ApplicationDbContext(DbContextOptions<ApplicationDbContext> options)
        : base(options) // Connection via DI, not hardcoded
    {
    }

    public DbSet<Product> Products => Set<Product>();
    public DbSet<Category> Categories => Set<Category>();
    public DbSet<Review> Reviews => Set<Review>();
    public DbSet<Tag> Tags => Set<Tag>();
    public DbSet<ProductTag> ProductTags => Set<ProductTag>();

    protected override void OnModelCreating(ModelBuilder modelBuilder)
    {
        base.OnModelCreating(modelBuilder);

        // Product configuration
        modelBuilder.Entity<Product>(entity =>
        {
            entity.HasKey(e => e.Id);
            
            entity.HasIndex(e => e.Slug).IsUnique();
            
            // Global query filter for soft delete (EF Core feature!)
            entity.HasQueryFilter(e => e.DeletedAt == null);

            // One-to-many relationship
            entity.HasOne(e => e.Category)
                .WithMany(c => c.Products)
                .HasForeignKey(e => e.CategoryId)
                .OnDelete(DeleteBehavior.Restrict);

            // One-to-many
            entity.HasMany(e => e.Reviews)
                .WithOne(r => r.Product)
                .HasForeignKey(r => r.ProductId);
        });

        // Many-to-many via join entity
        modelBuilder.Entity<ProductTag>(entity =>
        {
            entity.HasKey(pt => new { pt.ProductId, pt.TagId });

            entity.HasOne(pt => pt.Product)
                .WithMany(p => p.ProductTags)
                .HasForeignKey(pt => pt.ProductId);

            entity.HasOne(pt => pt.Tag)
                .WithMany(t => t.ProductTags)
                .HasForeignKey(pt => pt.TagId);
        });
    }

    // Override SaveChangesAsync for timestamps (async version)
    public override Task<int> SaveChangesAsync(CancellationToken cancellationToken = default)
    {
        var entries = ChangeTracker.Entries<Product>();

        foreach (var entry in entries)
        {
            if (entry.State == EntityState.Added)
            {
                entry.Entity.CreatedAt = DateTime.UtcNow; // Use UTC
            }
            else if (entry.State == EntityState.Modified)
            {
                entry.Entity.UpdatedAt = DateTime.UtcNow;
            }
        }

        return base.SaveChangesAsync(cancellationToken);
    }
}

// -----------------------------------------------------------------------------
// Query Extensions for Reusable Filters (Extensions/ProductQueryExtensions.cs)
// Replaces repository methods with composable LINQ extensions
// -----------------------------------------------------------------------------

namespace ProjectName.Extensions;

public static class ProductQueryExtensions
{
    public static IQueryable<Product> Active(this IQueryable<Product> query)
    {
        return query.Where(p => p.IsActive);
    }

    public static IQueryable<Product> InCategory(this IQueryable<Product> query, int categoryId)
    {
        return query.Where(p => p.CategoryId == categoryId);
    }

    public static IQueryable<Product> PriceRange(this IQueryable<Product> query, decimal min, decimal max)
    {
        return query.Where(p => p.Price >= min && p.Price <= max);
    }
    
    public static IQueryable<Product> IncludeSoftDeleted(this IQueryable<Product> query)
    {
        return query.IgnoreQueryFilters();
    }
}

// Usage:
// var products = await _context.Products
//     .Active()
//     .InCategory(5)
//     .PriceRange(10, 100)
//     .ToListAsync();

// -----------------------------------------------------------------------------
// KEY CONVERSION NOTES:
// -----------------------------------------------------------------------------
// 1. DbContext constructor: connection string → DbContextOptions via DI
// 2. virtual keyword not required for navigation properties
// 3. Many-to-many: implicit join table → explicit join entity (ProductTag)
// 4. Soft delete: manual filtering → HasQueryFilter (automatic!)
// 5. StringLength → MaxLength attribute
// 6. Collections: HashSet → List (or keep HashSet)
// 7. Nullable reference types: string → string? for optional
// 8. DateTime.Now → DateTime.UtcNow for cloud compatibility
// 9. SaveChanges → SaveChangesAsync (always async)
// 10. DbModelBuilder → ModelBuilder (Fluent API similar but different)
// 11. LazyLoadingEnabled = false → Default in EF Core (explicit Include)
// 12. Simple getters can use [NotMapped] computed properties
